### PR TITLE
MOD-16172 - support exclude empty in mrange commands

### PR DESCRIPTION
**Pull request by TalBarYakar**
> [!NOTE]
> **Medium Risk**
> Changes multi-key range query reply shape and cluster merge behavior; incorrect filtering or reply lengths could break clients relying on MRANGE semantics.
> 
> **Overview**
> Adds optional **`EXCLUDEEMPTY`** (since 8.10.0) to **`TS.MRANGE`** and **`TS.MREVRANGE`**, documented in `commands.json` and command info.
> 
> When set, matching series that would return **no samples** for the query (range, filters, aggregation, etc.) are **omitted** from the multi-series reply instead of appearing with empty sample arrays. Parsing stores the flag on **`MRangeArgs`**, rejects **`EXCLUDEEMPTY` with `GROUPBY`**, and avoids treating a **`GROUPBY`** label literally named `EXCLUDEEMPTY` as the flag.
> 
> Implementation probes each series with **`SeriesQueryIfNonEmpty`** and reuses that iterator via **`ReplySeriesRangeFromIter`** so empty series are skipped without a second scan. Top-level array/map length uses **postponed length** where needed so RESP2/RESP3 counts match the filtered series list.
> 
> The flag is propagated for **cluster/LibMR** (`QueryPredicates_Arg` serialize/deserialize and coordinator shard args). Extensive flow tests cover combinations with labels, count, aggregation, filter-by-value/ts, MREVRANGE, RESP3, and EMPTY aggregation edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 2cc7c0680c6106be9559ed3b2e19929adc36b6e7. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/reply.c [hunk 1]
```diff
@@ -129,6 +129,16 @@ int ReplySeriesArrayPos(RedisModuleCtx *ctx,
     return REDISMODULE_OK;
 }
 
+bool SeriesHasSamplesInRange(Series *series, const RangeArgs *args, bool reverse) {
+    RangeArgs oneArgs = *args;
+    oneArgs.count = 1;
```
Comment (AvivDavid23) [inline src/reply.c]: can explain? what setting it to one should do?

File: src/reply.c [hunk 2]
```diff
@@ -129,6 +129,16 @@ int ReplySeriesArrayPos(RedisModuleCtx *ctx,
...
+    AbstractIterator *iter = SeriesQuery(series, &oneArgs, reverse, true);
```
Comment (gabsow) [inline src/reply.c]: **EXCLUDEEMPTY + `AGGREGATION … EMPTY` is a bucket check, not a raw-sample check — and inconsistently so.**

This probe runs the full pipeline, including the aggregation iterator. With the `EMPTY` flag, `agg_iter_on_empty_chunk` fills NaN buckets across the whole range; `fillEmptyBuckets`' edge-gap check (`filter_iterator.c:742`) only *drops* them when there is no neighbor on one side. So a series with **zero raw samples in `[start,end]` but data on both sides** gets NaN buckets → `num_samples>0` → **kept**, while the same "no raw samples" series with data on only one side is **dropped**.

Repro: `TS.ADD s 10 1; TS.ADD s 9999 1` then `TS.MRANGE 1000 5000 AGGREGATION avg 100 EMPTY EXCLUDEEMPTY FILTER ...` returns `s` (all-NaN), but querying `[0,99]` excludes it. This contradicts the function name and the docstring of `test_excludeempty_agg_empty_flag_no_raw_samples` ("raw-sample check, not bucket check"). It is self-consistent (the reply would also contain those buckets), so arguably WAI — but please pick a semantic and document it, or drop aggregation from the probe to make it truly raw-sample.

File: src/reply.c (hunk 1 above)
Comment (gabsow) [inline src/reply.c]: **`oneArgs.count = 1` is a no-op**, and the probe is a second full query.

`SeriesQuery` (`tsdb.c:1363`) never reads `RangeArgs.count` — the count limit is applied only by the reply consumer in `ReplySeriesRange` (`reply.c:146`). So this line doesn't bound the work: `GetNext` still materializes a full first `EnrichedChunk` (and for aggregation may scan a whole bucket span). It also means that for every *kept* (non-empty) series the entire filter/aggregation pipeline runs **twice** — once here, once in `ReplySeriesRange`. Either drop the misleading line, or wire a real short-circuit. (The streaming reply emits key/labels before samples, so fully avoiding the pre-pass is hard — but worth a comment acknowledging the cost.)

File: src/query_language.c [hunk 3]
```diff
@@ -991,6 +991,8 @@ int parseMRangeCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc,
     }
     args.queryPredicates = queries;
 
+    args.excludeEmpty = RMUtil_ArgIndex("EXCLUDEEMPTY", argv, argc) > 0;
```
Comment (gabsow) [inline src/query_language.c]: **`RMUtil_ArgIndex` matches the token anywhere in argv** (full-token, case-insensitive — `deps/RedisModulesSDK/rmutil/util.c:35`), so a label literally named `EXCLUDEEMPTY` toggles the flag.

Concrete: `TS.MRANGE 1 100 FILTER l=v GROUPBY EXCLUDEEMPTY REDUCE sum` — grouping by a label named `EXCLUDEEMPTY` sets `excludeEmpty=true`, which then trips the new `EXCLUDEEMPTY is not allowed with GROUPBY` error below and rejects a valid query. Adding it to `QUERY_TOKENS` also means `SELECTED_LABELS EXCLUDEEMPTY` can no longer select that label. Edge-case and consistent with how sibling flags are parsed, but it changes reply *membership*, not just formatting — worth a note in docs at minimum.

**Review by gabsow — COMMENTED** [review]
Automated review (`/code-review`, extra-high effort). The feature is solid — postponed-length accounting is balanced on all paths, key/string cleanup in the new skip branch is correct, and `excludeEmpty` is initialized at every `MRangeArgs` construction site. No crash/leak/protocol bugs. The design choice to run the *same* `SeriesQuery` pipeline for the emptiness probe as for the reply is good: it guarantees "kept ⟺ reply non-empty". The inline comments below cover one behavioral inconsistency (EMPTY interaction), a parsing edge, two test-robustness gaps, and a few cleanup/altitude items. None are blocking.

File: src/query_language.c [hunk 4]
```diff
@@ -1017,6 +1019,12 @@ int parseMRangeCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc,
             goto error_free_all;
         }
     }
+
+    if (args.excludeEmpty && args.groupByLabel != NULL) {
+        RTS_ReplyGeneralError(ctx, "TSDB: EXCLUDEEMPTY is not allowed with GROUPBY");
```
Comment (gabsow) [inline src/query_language.c]: _Altitude (non-blocking):_ `EXCLUDEEMPTY` is a post-filter on the reply path that has to *forbid* `GROUPBY` rather than support it, and the same exclude+count logic is then re-implemented at three reply sites. A more general "suppress series/groups that emitted no rows" decision in the result-set/reply layer would apply uniformly to grouped and ungrouped output and avoid this special-case rejection. Reasonable as an MVP scope — just flagging the deeper shape.

File: src/libmr_commands.c [hunk 5]
```diff
@@ -183,20 +186,29 @@ static void mrange_done_internal(ExecutionCtx *eCtx, RedisModuleCtx *ctx, MRange
...
-                                    s,
-                                    args->withLabels,
-                                    args->limitLabels,
-                                    args->numLimitLabels,
-                                    &args->rangeArgs,
-                                    args->reverse,
-                                    false);
+            } else {
+                if (!args->excludeEmpty ||
+                    SeriesHasSamplesInRange(s, &args->rangeArgs, args->reverse)) {
```
Comment (gabsow) [inline src/libmr_commands.c]: **Duplication:** the `if (!excludeEmpty || SeriesHasSamplesInRange(...))` + `replylen++` + postponed-`ReplySetMapOrArrayLength` pattern is hand-rolled three times — here, in the near-identical `mrange_done_gears` (line 303), and in `replyUngroupedMultiRange` (`module.c:476`). A fix to the emit logic must land in all three and stay consistent across standalone vs. clustered topologies. Consider a shared `emitSeriesList(ctx, series_list, args, &replylen)` helper.

File: tests/flow/test_ts_mrange.py [hunk 6]
```diff
@@ -586,3 +586,398 @@ def test_mrange_nan_handling():
...
+        res = r1.execute_command(
+            'TS.MRANGE', 0, 299, 'AGGREGATION', 'avg', 100, 'EXCLUDEEMPTY', 'FILTER', 'g=pb')
+        assert len(res) == 1
+        assert res[0][0] == b'pb1'
+        # only one bucket has a value
+        samples = res[0][2]
+        assert any(s[1] == b'42' for s in samples)
+
+
+def test_excludeempty_agg_empty_flag_no_raw_samples(env):
```
Comment (gabsow) [inline tests/flow/test_ts_mrange.py]: This test passes **by accident** and doesn't guard its stated intent. `ef2`'s out-of-range sample is at `9999` (right side only), so the left-neighbor lookup returns 0 and the NaN buckets are dropped — exclusion happens for the wrong reason. The docstring claims "raw-sample check, not bucket check", but the bracketed case (an out-of-range sample on *each* side, see the `reply.c` comment on the `EMPTY` interaction) is never exercised and would make `assert len(res) == 1` fail. Please add a left-side out-of-range sample (and ideally a cluster variant) to actually cover it.

File: tests/flow/test_ts_mrange.py [hunk 7]
```diff
@@ -586,3 +586,398 @@ def test_mrange_nan_handling():
...
+        r.execute_command('TS.ADD', 'xa2', 5000, 1.0)
+
+        res = r1.execute_command(
+            'TS.MRANGE', 0, 99, 'AGGREGATION', 'sum', 50, 'EXCLUDEEMPTY', 'FILTER', 'g=agg1')
+        assert len(res) == 1
+        assert res[0][0] == b'xa1'
+        # two buckets: [0-49] sum=0+10+20+30+40=100, [50-99] sum=50+60+70+80+90=350
+        samples = res[0][2]
+        assert len(samples) == 2
+        assert samples[0] == [0, b'100']
```
Comment (gabsow) [inline tests/flow/test_ts_mrange.py]: These assertions hardcode the **RESP2** reply shape (`samples[0] == [0, b'100']`, and elsewhere `res[0][0]` / `dict(res[0][1])`). Under RESP3 the ungrouped MRANGE reply is a *map* and aggregation values come back as doubles, so the whole `EXCLUDEEMPTY` suite breaks. Sibling tests in this file guard with `is_resp3`; these use the bare `env` fixture with no protocol guard. Latent until the suite runs under `--protocol 3`.

File: tests/flow/test_ts_mrange.py [hunk 8]
```diff
@@ -586,3 +586,398 @@ def test_mrange_nan_handling():
...
+        assert samples[1] == [50, b'350']
+
+
+def test_excludeempty_agg_all_types(env):
+    """EXCLUDEEMPTY with every supported aggregation type — only the non-empty series returned."""
+    with env.getClusterConnectionIfNeeded() as r, env.getConnection(1) as r1:
+        for agg_type in ['avg', 'sum', 'min', 'max', 'count', 'first', 'last']:
+            env.flush()
```
Comment (gabsow) [inline tests/flow/test_ts_mrange.py]: _Test robustness:_ `env.flush()` inside the per-agg-type loop (and the fixed key names `at1`/`at2` reused each iteration) couples iterations through global state; most of the other new tests reuse fixed keys with no teardown at all. This is fragile under `--env-reuse` (a known sore spot in this repo's flow suite). Prefer `@pytest.mark.parametrize` over the agg types or unique key suffixes per case.

File: src/reply.c (hunk 1 above)
Comment (TalBarYakar) [reply src/reply.c]: not relevant anymore , removed

File: src/reply.c (hunk 2 above)
Comment (TalBarYakar) [reply src/reply.c]: if the user explicitly requested AGGREGATION … EMPTY, NaN buckets are part of the reply and EXCLUDEEMPTY should not override that in my opinion ... I waiting for lior's opinion as well

File: src/reply.c (hunk 1 above)
Comment (TalBarYakar) [reply src/reply.c]: removed

File: src/query_language.c (hunk 3 above)
Comment (TalBarYakar) [reply src/query_language.c]: This is consistent with how all other flags are parsed in this file (WITHLABELS, COUNT, etc.) — RMUtil_ArgIndex is the established pattern. A label named EXCLUDEEMPTY is an extreme edge case that applies equally to all existing tokens

File: src/query_language.c (hunk 4 above)
Comment (TalBarYakar) [reply src/query_language.c]: Acknowledged as a future refactor opportunity. The GROUPBY restriction is spec-mandated for now. Agreed the result-set layer would be the right place to generalize this eventually.

File: src/libmr_commands.c (hunk 5 above)
Comment (TalBarYakar) [reply src/libmr_commands.c]: In the new implementation filtering is done per inner shard in replyUngroupedMultiRange only, so this duplication doesn't arise.

File: tests/flow/test_ts_mrange.py (hunk 6 above)
Comment (TalBarYakar) [reply tests/flow/test_ts_mrange.py]: The semantic is now defined as reply-level (NaN buckets count as non-empty). The test will be updated to cover the both-sides case and assert the series is kept, not excluded.

File: tests/flow/test_ts_mrange.py (hunk 7 above)
Comment (TalBarYakar) [reply tests/flow/test_ts_mrange.py]: The new tests follow the same pattern as sibling tests in this file — RESP2 assertions in the main suite, RESP3 covered by a dedicated test_excludeempty_resp3 function that guards with is_resp3_possible and runs Env(protocol=3). The samples[0] == [0, b'100'] assertion is in test_excludeempty_agg_basic which only runs under RESP2.

File: tests/flow/test_ts_mrange.py (hunk 8 above)
Comment (TalBarYakar) [reply tests/flow/test_ts_mrange.py]: eplaced the loop with individual test functions per agg type (test_excludeempty_agg_type_avg, _sum, etc.), each using unique key names (xeat1_{agg_type}). No env.flush(), no shared state between iterations. RLTest doesn't support @pytest.mark.parametrize so the per-function expansion is the equivalent approach here.

File: src/reply.c [hunk 9]
```diff
@@ -136,6 +136,16 @@ int ReplySeriesArrayPos(RedisModuleCtx *ctx,
...
+    EnrichedChunk *chunk = iter->GetNext(iter);
+    bool hasData = chunk && chunk->samples.num_samples > 0;
+    iter->Close(iter);
+    return hasData;
```
[the remark this answers is not in the document]
Comment (TalBarYakar) [reply src/reply.c]: EXCLUDEEMPTY filters at the reply level, not raw-sample level. NaN buckets from AGGREGATION … EMPTY mean the series produced output — it shouldn't be excluded. This is intentional and covered by tests (test_excludeempty_agg_empty_flag_both_sides, test_excludeempty_agg_empty_flag_one_side).

File: src/module.c [hunk 10]
```diff
@@ -463,6 +463,12 @@ int replyUngroupedMultiRange(RedisModuleCtx *ctx, RedisModuleDict *result, const
             continue;
         }
 
+        if (args->excludeEmpty &&
+            !SeriesHasSamplesInRange(series, &args->rangeArgs, args->reverse)) {
+            RedisModule_CloseKey(key);
+            RedisModule_FreeString(ctx, currentKey);
+            continue;
+        }
```
[the remark this answers is not in the document]
Comment (TalBarYakar) [reply src/module.c:475]: EXCLUDEEMPTY filters at the reply level, not raw-sample level. NaN buckets from AGGREGATION … EMPTY mean the series produced output — it shouldn't be excluded. This is intentional and covered by tests (test_excludeempty_agg_empty_flag_both_sides, test_excludeempty_agg_empty_flag_one_side).

File: src/query_language.c [hunk 11]
```diff
@@ -1017,6 +1019,12 @@ int parseMRangeCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc,
...
+        goto error_free_all;
```
[the remark this answers is not in the document]
Comment (TalBarYakar) [reply src/query_language.c]: This is consistent with existing parsing — all flags in parseMRangeCommand use RMUtil_ArgIndex over the full argv (e.g. WITHLABELS, REVERSE). A label literally named EXCLUDEEMPTY would hit the same issue with any other flag name. WAI.

File: src/libmr_commands.c [hunk 12]
```diff
@@ -308,6 +310,15 @@ static void mrange_done_gears(ExecutionCtx *eCtx, RedisModuleCtx *ctx, MRangeDat
             Series *s = SeriesRecord_IntoSeries((SeriesRecord *)raw_record);
             tempSeries = array_append(tempSeries, s);
 
+            EnrichedChunk *first_chunk = NULL;
+            AbstractIterator *probe = NULL;
+            if (data->args.excludeEmpty) {
+                probe = SeriesQueryIfNonEmpty(
+                    s, &data->args.rangeArgs, data->args.reverse, &first_chunk);
+                if (!probe)
+                    continue;
```
Comment (gabsow) [inline src/libmr_commands.c:311]: **Critical: RESP array/map length was declared before this filtering runs — protocol corruption.**

A few lines above (line 290), `total_len` is computed by summing `ListRecord_GetLen(...)` over all shard results — the *unfiltered* series count — and `ReplyWithMapOrArray(ctx, total_len, false)` commits to that as a fixed (non-postponed) array/map length. This `continue` then skips `ReplySeriesArrayPos` entirely for any excluded series, without adjusting that already-declared length.

Any series skipped here means fewer elements are actually written than the declared length, corrupting the RESP reply framing for the connection (the client will hang waiting for more elements, or misparse the next reply).

Contrast with `replyUngroupedMultiRange` in `src/module.c`, which gets this right via `REDISMODULE_POSTPONED_ARRAY_LEN` + an incrementing `replylen` + `ReplySetMapOrArrayLength` at the end.

Repro: run a cluster configured with `ts-libmr-protocol GEARS` (supported, non-default) and issue `TS.MRANGE 1 100 EXCLUDEEMPTY FILTER g=x` where at least one matching series has no samples in range — the coordinator will declare more elements than it writes.

Fix: switch this branch to the same postponed-length pattern instead of pre-computing `total_len` up front.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/query_language.c (hunk 3 above)
Comment (gabsow) [inline src/query_language.c]: **`RMUtil_ArgIndex("EXCLUDEEMPTY", argv, argc)` scans the entire, unscoped argv — this can misfire against a user-supplied GROUPBY label.**

A few lines below, `args.groupByLabel = RedisModule_StringPtrLen(argv[groupby_location + 1], NULL)` takes the GROUPBY argument as a raw, unrestricted user string. Since this line scans the *whole* `argv`/`argc` (not just the "options" region), a time series with a label literally named `EXCLUDEEMPTY` used as:

```
TS.MRANGE 1 100 FILTER somelabel=val GROUPBY EXCLUDEEMPTY REDUCE sum
```

will make this line find that GROUPBY-label token and set `args.excludeEmpty = true`, even though the user never passed the EXCLUDEEMPTY flag. That then trips the new guard below (`if (args.excludeEmpty && args.groupByLabel != NULL) { RTS_ReplyGeneralError(ctx, "TSDB: EXCLUDEEMPTY is not allowed with GROUPBY"); }`), rejecting a previously-valid command with a misleading error, purely because of the label's name.

(FILTER label values are safe from this collision since `parseLabelListFromArgs` requires every filter token to contain `=`, so a bare `EXCLUDEEMPTY` token can only come from the unconstrained GROUPBY-label slot — or `SELECTED_LABELS`, though that loop already stops at reserved tokens including this new one.)

Suggest scoping this lookup to before `groupby_location` (mirroring how `last_filter_pos`/`query_count` are already computed relative to `groupby_location`), or checking the flag only among the trailing options after the FILTER predicate list ends.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/libmr_commands.c [hunk 13]
```diff
@@ -308,6 +310,15 @@ static void mrange_done_gears(ExecutionCtx *eCtx, RedisModuleCtx *ctx, MRangeDat
...
+            }
+
             if (data->args.groupByLabel) {
```
Comment (gabsow) [inline src/libmr_commands.c:314]: **Latent iterator leak if `excludeEmpty` and `groupByLabel` are ever both true.**

The probe block above (lines 315-320) runs unconditionally whenever `data->args.excludeEmpty` is set, *before* this `groupByLabel` branch. If this branch is taken with a live `probe` (non-NULL `AbstractIterator*`), it's simply discarded here — `ResultSet_AddSeries(resultset, s, ...)` doesn't take or close it, so `probe` (and the iterator chain it holds open) leaks.

Today this is unreachable in practice only because `parseMRangeCommand` (src/query_language.c) rejects `EXCLUDEEMPTY + GROUPBY` at parse time — an invariant enforced in a different file, far from this code. Nothing here locally asserts or structurally prevents the combination, so a future caller that constructs `MRangeArgs` another way (or a refactor that relaxes the parser-level guard) would silently leak an iterator per kept series with no crash or error to signal it.

Worth guarding this probe with `data->args.excludeEmpty && !data->args.groupByLabel` (or asserting the invariant) so the safety doesn't depend entirely on a distant, easy-to-forget guard.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/module.c [hunk 14]
```diff
@@ -463,14 +463,26 @@ int replyUngroupedMultiRange(RedisModuleCtx *ctx, RedisModuleDict *result, const
             continue;
         }
 
+        EnrichedChunk *first_chunk = NULL;
+        AbstractIterator *probe = NULL;
+        if (args->excludeEmpty) {
+            probe = SeriesQueryIfNonEmpty(series, &args->rangeArgs, args->reverse, &first_chunk);
```
Comment (gabsow) [inline src/module.c:469]: **This probe-and-skip block is duplicated near-verbatim in `src/libmr_commands.c`'s `mrange_done_gears` (the `if (data->args.excludeEmpty) { probe = SeriesQueryIfNonEmpty(...); if (!probe) continue; }` block), with no shared helper.**

Both copies thread `probe`/`first_chunk` through to `ReplySeriesArrayPos` and each does its own cleanup on the empty branch (this one closes the key/frees the string; the other just `continue`s). Because the logic isn't centralized, it's easy for a call site to miss it entirely — which is exactly what happened one call further out: `resultset.c`'s `GroupList_ReplyResultSet` and `mrange_done_gears`'s own groupBy branch both always pass `NULL, NULL` and rely on the GROUPBY+EXCLUDEEMPTY rejection living in yet another file (`query_language.c`) to stay correct (see the related comment on `libmr_commands.c`).

Consider extracting a single helper (e.g. `bool ExcludeEmptyShouldSkip(Series*, const RangeArgs*, bool reverse, AbstractIterator**, EnrichedChunk**)`, or folding the probe directly into `ReplySeriesArrayPos` given an `excludeEmpty` flag) so every current and future call site inherits the same behavior instead of re-implementing it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/query_language.c [hunk 15]
```diff
@@ -1017,6 +1019,12 @@ int parseMRangeCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc,
             goto error_free_all;
         }
     }
+
+    if (args.excludeEmpty && args.groupByLabel != NULL) {
```
Comment (gabsow) [inline src/query_language.c]: Minor: this rejection runs after `parseFilter` has already built the full `QueryPredicateList` (label parsing + predicate-type counting), and — when GROUPBY is present — after `parseMultiSeriesReduceArgs` has already parsed/allocated the reducer args too. A guaranteed-to-error `EXCLUDEEMPTY ... GROUPBY ...` request currently pays for all of that work first. Moving this check right after `args.excludeEmpty` is computed (before `parseFilter` runs) would fail fast. Low severity since it's parse-time cost on an already-invalid request, not a per-series cost.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by gabsow — COMMENTED** [review]
Automated code review (high effort, 8 finder angles + verification passes). Summary of findings below; full detail in the inline comments.

**Critical:** `mrange_done_gears` (src/libmr_commands.c) declares the RESP array/map reply length from the unfiltered series count, then skips series via `EXCLUDEEMPTY`'s `continue` without adjusting that length — corrupting the reply when running with `ts-libmr-protocol GEARS` and at least one excluded series. `mrange_done_internal` (the default `LIBMR_PROTOCOL_INTERNAL` path) does **not** have this problem — verified that shard-side `replyUngroupedMultiRange` already filters correctly with a postponed length before the coordinator ever sees the data.

**High:** The new `EXCLUDEEMPTY` flag is detected via an unscoped `RMUtil_ArgIndex` scan of the full `argv` in `parseMRangeCommand` (src/query_language.c), which can collide with a user-supplied `GROUPBY <label>` name literally equal to `EXCLUDEEMPTY`, incorrectly triggering the new "not allowed with GROUPBY" rejection on an otherwise valid command.

**Medium:** A latent iterator leak in `mrange_done_gears` if `excludeEmpty` and `groupByLabel` are ever both true (currently prevented only by a parse-time guard in a different file), and duplicated probe-and-skip logic across `module.c`/`libmr_commands.c` with no shared helper — which is exactly how the reply-length bug and the leak risk both crept in.

**Low:** minor parse-order inefficiency when EXCLUDEEMPTY+GROUPBY are combined.

This review was generated with multiple independent finder agents plus targeted verification passes tracing the actual call graph (LIBMR_PROTOCOL_INTERNAL vs GEARS dispatch, `SeriesListReplyParser`, `ResultSet_AddSeries` signature, etc.) to separate real bugs from false leads.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/libmr_commands.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (TalBarYakar) [reply src/libmr_commands.c:311]: Fixed in the current implementation. The LibMR Gears ungrouped path no longer declares the full shard series count up front. It now uses REDISMODULE_POSTPONED_ARRAY_LEN, increments replylen only for series that are actually emitted after EXCLUDEEMPTY filtering, and finalizes the reply with ReplySetMapOrArrayLength(ctx, replylen, false). So skipped empty series no longer corrupt the MRANGE/MREVRANGE reply length.

Comment (TalBarYakar) [reply src/libmr_commands.c:311]: Fixed. The Gears ungrouped path now follows the same postponed-length pattern as `replyUngroupedMultiRange`: it replies with `REDISMODULE_POSTPONED_ARRAY_LEN`, increments `replylen` only after `ReplySeriesArrayPos` actually emits a series, and calls `ReplySetMapOrArrayLength(ctx, replylen, false)` after the loop. So an `EXCLUDEEMPTY` skip no longer leaves the RESP array/map length larger than the number of emitted entries.

File: src/query_language.c (hunk 3 above)
Comment (TalBarYakar) [reply src/query_language.c]: Fixed. The parser now keeps the position returned by `RMUtil_ArgIndex("EXCLUDEEMPTY", argv, argc)` and ignores it when that token is exactly the `GROUPBY` label argument. So `GROUPBY EXCLUDEEMPTY REDUCE ...` is treated as a label named `EXCLUDEEMPTY`, not as the `EXCLUDEEMPTY` flag, and it no longer trips the `EXCLUDEEMPTY is not allowed with GROUPBY` guard.

File: src/libmr_commands.c (hunk 13 above)
Comment (TalBarYakar) [reply src/libmr_commands.c:314]: I’m going to leave this as-is. `EXCLUDEEMPTY + GROUPBY` is rejected in `parseMRangeCommand`, so `mrange_done_gears` should not receive `data->args.excludeEmpty && data->args.groupByLabel` from the command path. The probe is only reachable for valid ungrouped EXCLUDEEMPTY requests, and adding `&& !data->args.groupByLabel` here would duplicate the parser invariant rather than fix a reachable leak.

File: src/module.c (hunk 14 above)
Comment (TalBarYakar) [reply src/module.c:469]: I’m going to keep this local for now. The shared operation is already factored into `SeriesQueryIfNonEmpty(...)`, and the remaining control flow is call-site-specific: `replyUngroupedMultiRange` must close the Redis key and free `currentKey` before continuing, while `mrange_done_gears` does not own those same resources in that loop. `ReplySeriesArrayPos` already accepts `probe`/`first_chunk`, so extracting another helper here would mostly wrap a small branch while still needing caller-specific cleanup.

File: src/query_language.c (hunk 15 above)
Comment (TalBarYakar) [reply src/query_language.c]: fixed

**Review by gabsow — APPROVED** [review]

**Review by gabsow — DISMISSED** [review]
Approving. Re-checked against the current head (`940cf2298`) — the blocking issues are all fixed:

- **Critical** — RESP reply-length mismatch (`libmr_commands.c:311`): the declared array/map length now matches the elements emitted on the exclude path. ✅
- **HIGH** — GEARS/cluster MRANGE now applies EXCLUDEEMPTY (`libmr_commands.c:643`). ✅
- Double-query-per-kept-series (`module.c:475`), EXCLUDEEMPTY parsing (`query_language.c:994`), and the reply shape (`reply.c:134`) — resolved. The earlier iterator-leak concern (`libmr_commands.c:314`) doesn't actually leak on re-check.

A few **low/nit** items remain (non-blocking — worth a quick follow-up):
- `SELECTED_LABELS` + `EXCLUDEEMPTY` on a label (`query_language.c:27`) — not yet handled.
- Minor cleanups around `query_language.c:994/1023` and `libmr_commands.c:193`.

LGTM.

**Review by AvivDavid23 — APPROVED** [review]
